### ISSUE TITLE
Add notebook support for Quarto and RMarkdown with Chunk Output inline

### ISSUE THREAD

**Issue by StefKirsch** [posit-dev/positron#5640]
This is a plea to bring back the amazing notebook support from RStudio in positron as well. I decided to create an issue despite the advice given [here](https://github.com/posit-dev/positron/discussions/3718#discussioncomment-9938783) because I only found it after I did several hours of research on this and only found it while typing this issue. Might be a me issue, but the discussions seem hard to find.

## Background

One amazing benefit RStudio has over VS Code and other IDE's is it's notebook support for other file types that .ipynb. As is explained in the RStudio documentation [here](https://rmarkdown.rstudio.com/lesson-10.html), Jupyter Notebooks are very cumbersome as soon as you start putting them under version control because
1. They are not plain text files, which make diffs difficult to handle and as far as I can tell basically impossible to selectively stage changes
2. The output is stored directly in the file, and to ignore those in version control you need further setup or third-party tools like nbstripout, both of which add complexity and are prone to errors.

This has in my experience effectively negated one of the big unique selling point of notebooks, which is the ease of access especially while you are still learning. As an educator, this is the biggest reason why I have stuck to R and RStudio for learners in academia.

## Current situation in Positron

As I see it now, it is not planned for Positron to support other notebook formats than .ipynb (see #2866) anytime soon. Also see this discussion [here](https://github.com/posit-dev/positron/discussions/3718#discussioncomment-9938783).

I assume there are reasons for this, which has probably a lot to do with the fact that .qmd and .rmd also don't have notebook support in VS Code. 

## Workaround

Of course, we can use the interactive mode and get the output next to the notebook or even work with the. I get why some people are fine with that or even prefer it (see #1533). But this will always split the code from the output and blow up the interface. Hence the need for notebooks in the first place.

## Expected or desired behavior

Getting back the Notebook support of .rmd and .qmd files with Chunk Output inline in Positron.

**Comment by ntluong95** [comment]
I strongly favor for this, because version control for `.ipynb` file is suck

**Comment by frederikziebell** [comment]
I can just second this. One of the killer features of RStudio is the notebook support of .Rmd and .qmd files. If you would bring Quarto inline notebook support to Positron, we would have one universal notebook format (.qmd files), with which we can do multi-language interactive analysis and reporting! Without this support, I will never switch to Positron because I do 90% of my work as interactive analysis in notebook chunks.

**Comment by venpopov** [comment]
I commented in other threads as well, but I wholeheartedly support and want inline output for Quarto notebooks. In addition to the version control issues (which can be solved by adding yet more complexity to the system), the editing experience of .qmd files is vastly superior to jupyter notebooks! In Visual Editing mode quarto notebooks now look almost identical to jupyter notebooks (at least in VSCode), inline output is the only thing missing. In-line output makes rapid prototyping and teaching much more pleasant as **you have access to the output of your code in the correct context** - next to the code that produced it! You don't have to keep multiple windows open - an editor and a preview browser and constantly switch back and forth between them. Regular output sends stdout to console, figures to a figure panel, etc, making for a very disjointed experience.

One of the key arguments against inline outputs in other threads were complications concerning the environment in a console and in a notebook. A straightforward solution is to have a single shared environment, and that's it.

**Comment by stefanocoretta** [comment]
I second all of the above! Many many excellent reasons for why inline output is a must (with the option of turning it off for those who don't use it).

**Comment by ntluong95** [comment]
There are many reasons while supporting in-line output for .qmd file in Positron
- First, python user are very much used to .ipynb file, they will never change to .qmd because they want to view and keep outputs persistently.
- R user from Rstudio will miss the feature a lot and they may not want to use Positron because of it
- Supporting in-line output for .qmd file will make Positron outstanding with VSCode

**Comment by fbob** [comment]
Any update on this so useful requested feature ?

**Comment by issactoast** [comment]
up up up! :)

**Comment by amcadie** [comment]
One more comment from the peanut gallery here.  I have been using Positron for the last couple weeks and I am very excited about the tool!  You've done an amazing job combining the best of VSCode and RStudio!  The addition of a BigQuery extension means I can do my entire workflow in one IDE.  The one sticking point for me has been the lack of in-line output for Quarto notebooks, which feels like a regression.  The text-based file format combined with notebook experience is a killer combo, and would be truly next level with the language-agnostic Quarto notebook on the customizable Positron platform.  

From my standpoint, the coupling/decoupling of the notebook environment with the console environment is not an issue, and I could adapt either way.  I find working in the console cumbersome for the most part now that I've gotten used to working in notebooks.  What I do very much appreciate is the Session/Environment view.  So if decoupling the notebook and console environments makes more sense or is easier to implement on your end, perhaps adding to the environment dropdown would be workable for everyone.  For example, you might see `console/R 4.4.0`, `report.qmd/R 4.4.0`, etc as options to select and see the objects defined in that environment.

Thanks for considering and for all the work on these great tools!

**Comment by reddylee** [comment]
This is a must-have feature for my workflow. Without it, I’d strongly prefer to stick with RStudio.

**Comment by chrishklee** [comment]
Agree with all of the above. Inline chunk output from .qmd or .rmd is an absolute must-have for me.

**Comment by tripartio** [comment]
I really, really want to love Positron. I'm trying so hard to make the switch from RStudio that when I got a spanking new data science laptop, I only installed Positron with no RStudio to force me to make the full switch. I almost made it, even building my R package fully through Positron. Then I switched projects and came to a deadend with Positron's lack of inline, instantaneous Quarto notebooks. That killed it for me. Now I'm back to RStudio. I want one IDE that does it all (at least for R) rather than having to hop between Positron and RStudio for different projects.

Please, please integrate this full functionality in Positron. Just about everything else is great.

**Comment by FredMountain** [comment]
Not an enhancement but an essential from my perspective. The advice @juliasilge to use .ipynb instead unfortunately doens't work for me as well for two major reasons. 

1. Working together with colleagues who are on R studio. It doesn't make sense for them to use .ipynb in rstudio.   
2. I want to be able to work on rmarkdown and qmd with available output formats (bookdown etc.) and still have inline outputs while working on it.

These issues really keep me from switching and advicing others to give it a try.

**Comment by Bott-binc** [comment]
Without this feature, I won't switch to Positron.  It is too important to my workflow.  Agreed, its an **essential** feature, not an enhancement.

**Comment by joranE** [comment]
Just adding another pebble to pile here: Positron is essentially a non-starter for me unless it supports inline output in Quarto (and/or RMarkdown).

**Comment by LorennMarque** [comment]
"Without this feature, I won't switch to Positron. It is too important to my workflow" Totally agree, i cant use positron without inline output in Quarto.

**Comment by Ido-deBaat** [comment]
Oh! If only I could get paged table inline output for Quarto and RMarkdown notebooks in Positron 🥹 

I’ve been installing and deinstalling Positron over and over again since one of the first public beta versions, yearning to switch: be an early adopter, advocate for it, discover it, customise the interface… But it was not to be. Without inline (paged table) output I really miss the ergonomics that came with RMarkdown notebooks and frankly, my daily work life sort of evolves around that simpel and direct connection between input and output that continuously lets me validate and compare my results.

It’s been on my mind a lot and I somehow couldn’t find this topic earlier. Glad I’m not the only one 😅

**Comment by kv9898** [comment]
Guys, I'm confused about one thing: whether we want 1) just inline outputs after chunks or 2) we are okay with a 'notebook' experience (like that for `.ipynb` files) in `.qmd` files which do not record outputs.

AFAICT 2) would be relatively easy (we just need to make another notebook contribution) but 1 would be hard, as it involves major changes to the editor ui.

If you guys are okay with 2 I think I can try putting up an extension.

**Comment by ntluong95** [comment]
Option 2 would be okay for me. It is similar to Quarto/Rmd experience in Rstudio. The output is not binded to the document, but I think they are storage somewhere in the local memory, that's why when you open it you see some outputs from previous session

**Comment by kv9898** [comment]
> Option 2 would be okay for me. It is similar to Quarto/Rmd experience in Rstudio. The output is not binded to the document, but I think they are storage somewhere in the local memory, that's why when you open it you see some outputs from previous session

what I meant was that, with option 2, the markdown parts of qmd also have to be in 'chunks' and will be displayed as html when not being edited - this is different from rstudio experience.

**Comment by ntluong95** [comment]
ah I see now. It is equivalent to the `Visual mode` I guess

**Comment by AxesAccess** [comment]
> what I meant was that, with option 2, the markdown parts of qmd also have to be in 'chunks' and will be displayed as html when not being edited - this is different from rstudio experience.

Do you mean that cell output will be written inline below the cell in the qmd/rmd source code?

**Comment by kv9898** [comment]
> > what I meant was that, with option 2, the markdown parts of qmd also have to be in 'chunks' and will be displayed as html when not being edited - this is different from rstudio experience.
> 
> Do you mean that cell output will be written inline below the cell in the qmd/rmd source code?

That's the same for options 1 AND 2. But option 2 will not display raw markdown  - it is considered a "markdown code cell" and would be rendered as html when you are not editing it.

**Comment by LorennMarque** [comment]
Please add a comment when inline output is added to positron. The same way as it works with QMD notebooks in R studio. Thanks! Really hope u can contribute with the extension

**Comment by AxesAccess** [comment]
> That's the same for options 1 AND 2. But option 2 will not display raw markdown - it is considered a "markdown code cell" and would be rendered as html when you are not editing it.

It's already implemented as Jupyter notebooks, isn't it?

I still cannot understand what retains Posit from implementing inline output in Positron?

**Comment by kv9898** [comment]
> > That's the same for options 1 AND 2. But option 2 will not display raw markdown - it is considered a "markdown code cell" and would be rendered as html when you are not editing it.
> 
> It's already implemented as Jupyter notebooks, isn't it?
> 
> I still cannot understand what retains Posit from implementing inline output in Positron?

They are developing Positron Notebooks which operate a bit different to Jupyter and maybe they want to support qmd through that

**Comment by tripartio** [comment]
> Guys, I'm confused about one thing: whether we want 1) just inline outputs after chunks or 2) we are okay with a 'notebook' experience (like that for `.ipynb` files) in `.qmd` files which do not record outputs.
> 
> AFAICT 2) would be relatively easy (we just need to make another notebook contribution) but 1 would be hard, as it involves major changes to the editor ui.
> 
> If you guys are okay with 2 I think I can try putting up an extension.

@kv9898  I don't understand the difference you mean to convey between these two options, and the discussion that followed your question doesn't help clarify things. Could you please show screenshots of examples of the two different options, perhaps from RStudio and Jupyter notebooks as examples of what you mean concretely?

I understand that this entire thread is asking for the RStudio experience, not the Jupyter experience, but I don't understand precisely what you mean by either of your two options.

**Comment by kv9898** [comment]
> > Guys, I'm confused about one thing: whether we want 1) just inline outputs after chunks or 2) we are okay with a 'notebook' experience (like that for `.ipynb` files) in `.qmd` files which do not record outputs.
> > 
> > AFAICT 2) would be relatively easy (we just need to make another notebook contribution) but 1 would be hard, as it involves major changes to the editor ui.
> > 
> > If you guys are okay with 2 I think I can try putting up an extension.
> 
> @kv9898  I don't understand the difference you mean to convey between these two options, and the discussion that followed your question doesn't help clarify things. Could you please show screenshots of examples of the two different options, perhaps from RStudio and Jupyter notebooks as examples of what you mean concretely?
> 
> I understand that this entire thread is asking for the RStudio experience, not the Jupyter experience, but I don't understand precisely what you mean by either of your two options.

The rstudio experience would be option 1 then... Option 2 is precisely jupyter.

**Comment by tripartio** [comment]
> The rstudio experience would be option 1 then... Option 2 is precisely jupyter.

If you read the thread carefully, I think you will see that we are asking for the RStudio experience. Positron already supports the Jupyter experience, which those who have commented and voted on this thread have overwhelmingly indicated is subpar to inline Quarto and RStudio notebooks. Could you please clarify in what way what you are proposing for Option 2 would give Positron new capabilities that it does not already have?

**Comment by kv9898** [comment]
> > The rstudio experience would be option 1 then... Option 2 is precisely jupyter.
> 
> If you read the thread carefully, I think you will see that we are asking for the RStudio experience. Positron already supports the Jupyter experience, which those who have commented and voted on this thread have overwhelmingly indicated is subpar to inline Quarto and RStudio notebooks. Could you please clarify in what way what you are proposing for Option 2 would give Positron new capabilities that it does not already have?

I *have read* the read carefully and suggested support for qmd support in a jupyter style as a TECHNICAL COMPROMISE as it's easier done that way (vscode already has notebook api). Jupyter notebooks for now do not support qmd files and .ipynb is not good for version control.

**Comment by olalobeja** [comment]
> Guys, I'm confused about one thing: whether we want 1) just inline outputs after chunks or 2) we are okay with a 'notebook' experience (like that for `.ipynb` files) in `.qmd` files which do not record outputs.
> 
> AFAICT 2) would be relatively easy (we just need to make another notebook contribution) but 1 would be hard, as it involves major changes to the editor ui.
> 
> If you guys are okay with 2 I think I can try putting up an extension.

Thanks for the efforts @kv9898  I think option 1 as you state it is preferred, just like in RStudio basically. Having "markdown" cells like in Jupyter notebooks via html is not what RStudio does, and a preferred route for Positron behaviour for .QMD files  in my use case and recommendation is to directly develop exactly in line interaction like in RStudio. If this feature ever comes to fruition...wow, maybe a real Nobel Peace Prize for this effort!!!

**Comment by xymaxim** [comment]
@kv9898, thanks for your efforts on this issue.

As like @tripartio, I'm not sure if I fully understand the difference between your options.

To make it more clear, let me share some screenshots. I would prefer to avoid using any reference to Jupyter since this issue is about Quarto/RMarkdown, but rather stick with mentions to (1) the current Positron and (2) the desired RStudio-like behaviors for inline code chunk output in Quarto documents.

1) So, here is the current Positron behavior:

<img width="1366" height="768" alt="Image" src="https://github.com/user-attachments/assets/da8b4fa3-6ac1-4c47-abe2-9723f187658f" />

2) The desired behavior would be similar to RStudio's "Chunk Output Inline" option:

<img width="1366" height="768" alt="Image" src="https://github.com/user-attachments/assets/3513b103-ce58-4bc1-b39e-b09d7de7829b" />

> > Option 2 would be okay for me. It is similar to Quarto/Rmd experience in Rstudio. The output is not binded to the document, but I think they are storage somewhere in the local memory, that's why when you open it you see some outputs from previous session
> 
> what I meant was that, with option 2, the markdown parts of qmd also have to be in 'chunks' and will be displayed as html when not being edited - this is different from rstudio experience.

Now, what I don't quite follow is your message about the Markdown support. Why we should imitate the Jupyter experience (the Markdown rendering) here? As I see, Positron supports both source and visual modes outside code cells (R, Python, etc). Is it due to using the VS Code's [Notebook API](https://code.visualstudio.com/api/extension-guides/notebook) for Option 2? 

So, in short, your Option 2 will bring a VS Code extension, while Option 1 would require changes to the Positron codebase. Am I right here?

**Comment by kv9898** [comment]
> @kv9898, thanks for your efforts on this issue.
> 
> As like @tripartio, I'm not sure if I fully understand the difference between your options.
> 
> To make it more clear, let me share some screenshots. I would prefer to avoid using any reference to Jupyter since this issue is about Quarto/RMarkdown, but rather stick with mentions to (1) the current Positron and (2) the desired RStudio-like behaviors for inline code chunk output in Quarto documents.
> 
> 1) So, here is the current Positron behavior:
> 
> <img width="1366" height="768" alt="Image" src="https://github.com/user-attachments/assets/da8b4fa3-6ac1-4c47-abe2-9723f187658f" />
> 
> 2) The desired behavior would be similar to RStudio's "Chunk Output Inline" option:
> 
> <img width="1366" height="768" alt="Image" src="https://github.com/user-attachments/assets/3513b103-ce58-4bc1-b39e-b09d7de7829b" />
> 
> > > Option 2 would be okay for me. It is similar to Quarto/Rmd experience in Rstudio. The output is not binded to the document, but I think they are storage somewhere in the local memory, that's why when you open it you see some outputs from previous session
> > 
> > what I meant was that, with option 2, the markdown parts of qmd also have to be in 'chunks' and will be displayed as html when not being edited - this is different from rstudio experience.
> 
> Now, what I don't quite follow is your message about the Markdown support. Why we should imitate the Jupyter experience (the Markdown rendering) here? As I see, Positron supports both source and visual modes outside code cells (R, Python, etc). Is it due to using the VS Code's [Notebook API](https://code.visualstudio.com/api/extension-guides/notebook) for Option 2? 
> 
> So, in short, your Option 2 will bring a VS Code extension, while Option 1 would require changes to the Positron codebase. Am I right here?
> 

Yes, Option 2 can be done via an extension and Option 1 involves signifcant changes to Positron's codebase related to workbench contributions. For Option 1, a UI component needs to be added, and it needs to intercept text, plot and html output from the console, which is non-trivial. I guess that's why the Posit team has left this for so long.

**Comment by tripartio** [comment]
> [kv9898] Yes, Option 2 can be done via an extension and Option 1 involves signifcant changes to Positron's codebase related to workbench contributions. For Option 1, a UI component needs to be added, and it needs to intercept text, plot and html output from the console, which is non-trivial. I guess that's why the Posit team has left this for so long.

@kv9898, thanks for your attempt to help provide a partial solution, but as @xymaxim indicated, your proposal #2 is still not clear. @xymaxim did a great job visually displaying your proposal #1 with screenshots, indicating what I think most of us are asking for. Could you please replicate similar screenshots to show what you intend by your option #2? I think that would help us better respond to how helpful it might be.

**Comment by frederikziebell** [comment]
> Yes, Option 2 can be done via an extension and Option 1 involves signifcant changes to Positron's codebase related to workbench contributions. For Option 1, a UI component needs to be added, and it needs to intercept text, plot and html output from the console, which is non-trivial. I guess that's why the Posit team has left this for so long.

@juliasilge Could you maybe please outline Posit's plans to implement this feature, if any exist already? This is not to put pressure on the core dev team but rather to understand if it makes sense for the community to start working on a solution (e.g. @kv9898 [offered to create an extension](https://github.com/posit-dev/positron/issues/5640#issuecomment-3476938398)). Also, there was a speculation that RStudio's notebook experience will be brought to Positron via "Positron Notebooks". Is that correct?

**Comment by grst** [comment]
Note that in vscode there is also a third option which is kind of the intermediate between the Rstudio experience (code, plots, console completely separate) and the notebook experience (code + output inline). 

And that is to send quarto chunks to a jupyter console, that renders plots inline. So no output will be shown in the qmd, but chunks and all outputs will be shown inline in a separate console window. 

<img width="3087" height="1958" alt="Image" src="https://github.com/user-attachments/assets/828631b5-93d4-4e5f-ba63-0239f29e9dad" />

Alas, this currently only works for python, but not for R, due to limitations of the vscode API. See also https://github.com/quarto-dev/quarto/issues/204. But if this were to be addressed on the vscode/positron side, it could also be an elegant solution.

**Comment by kv9898** [comment]
> > [kv9898] Yes, Option 2 can be done via an extension and Option 1 involves signifcant changes to Positron's codebase related to workbench contributions. For Option 1, a UI component needs to be added, and it needs to intercept text, plot and html output from the console, which is non-trivial. I guess that's why the Posit team has left this for so long.
> 
> [@kv9898](https://github.com/kv9898), thanks for your attempt to help provide a partial solution, but as [@xymaxim](https://github.com/xymaxim) indicated, your proposal [#2](https://github.com/posit-dev/positron/pull/2) is still not clear. [@xymaxim](https://github.com/xymaxim) did a great job visually displaying your proposal [#1](https://github.com/posit-dev/positron/pull/1) with screenshots, indicating what I think most of us are asking for. Could you please replicate similar screenshots to show what you intend by your option [#2](https://github.com/posit-dev/positron/pull/2)? I think that would help us better respond to how helpful it might be.

I cannot replicate something that is not yet available.

Never mind, I realized creating an extension that adds Jupyter-style Notebook support for qmd is not what most people find acceptable and inevitably involves loss of original quarto functionalities due to VS Code API limitations. Therefore, it does not seem to be worth the effort.

**Comment by LorennMarque** [comment]
Is there any progress on the inline code cell output? Im waiting for this feature to start using positron. Thanks!

**Comment by Akash-Ansari** [comment]
This is the most crucial feature of my RStudio workflow.

**Comment by kv9898** [comment]
<img width="1600" height="876" alt="Image" src="https://github.com/user-attachments/assets/5bb65bf2-6da7-4dc9-a609-150c77a46dab" />

Happy new year and welcome to 2026 guys. It seems that we can leverage VS Code's `editorInsets` feature to embed results into the editor *as an extension* (so what I have shown here is doable solely in the [Quarto](https://github.com/quarto-dev/quarto/) extension). The real problem is that Positron is *not* passing plot/html data back through the `positron.runtime.ExecutionObserver` API, despite its doc says it should do so:

```ts
			/**
			 * An optional callback to invoke when the execution emits a plot.
			 *
			 * NOTE: Currently only fired for static plots, not dynamic plots.
			 *
			 * @param plotData The plot data emitted, as a string.
			 */
			onPlot?: (plotData: string) => void;
```

<img width="1295" height="45" alt="Image" src="https://github.com/user-attachments/assets/8b597c11-ee36-43cf-9952-081a2a064183" />

In the second debugging line, I'm actually doing a ggplot (static) and the plot is displayed normally in the plot pane, but I'm not seeing any plot data passed back. Is this a bug, or a feature that is not yet implemented?

**Comment by frederikziebell** [comment]
I had a very constructive [feedback meeting](https://scheduler.zoom.us/cindy-tong/improving-the-positron-notebook-experience) last week with @cindyytong and an engineer in Positron's new notebook team. I reiterated the need for plaintext notebook support for interactive analysis with inline chunk output, and also pointed to @clauswilke's [excellent suggestion](https://github.com/posit-dev/positron/issues/1533#issuecomment-2204416511) on how inline vs console output could be implemented.

I can't go into details or timeline, but based on that conversation I'm cautiously optimistic that inline chunk output, and more generally better support for interactive analysis with Quarto notebooks, will come. I came away with the impression that the team is *not* trying to push users toward Jupyter notebooks for interactive analysis, but rather that Jupyter was the initial focus for support and improvement, with Quarto now following.

**Comment by kaniastefan** [comment]
Just to keep up the momentum on this. For me this is the main sticking point why I have not changed from RStudio to Positron. It might seem like a small issue, but not having inline output is really disrupting my usual workflow. I don't like to use Jupyter as I cannot just run highlighted code in a cell.

**Comment by LorennMarque** [comment]
> Just to keep up the momentum on this. For me this is the main sticking point why I have not changed from RStudio to Positron. It might seem like a small issue, but not having inline output is really disrupting my usual workflow. I don't like to use Jupyter as I cannot just run highlighted code in a cell.

Agree! Im still waiting for this feature to start using positron

**Comment by cccfran** [comment]
Love the momentum and to see this improvement on the timeline! This feature is important for my research workflows and teaching. I will switch to Positron for sure once it lands. The extension ecosystem there is just amazing.

**Comment by kv9898** [comment]
<img width="1237" height="882" alt="Image" src="https://github.com/user-attachments/assets/e724c670-2415-454b-8c7f-f4848c0ba687" />

<img width="1238" height="173" alt="Image" src="https://github.com/user-attachments/assets/8a10dbc3-9fc1-4950-9de8-bc275ef5a4c1" />

The progress looks super promising! Thanks for the effort!!!

**Comment by AlbertoFabbri93** [comment]
Is this in the preview or stable version?

**Comment by kv9898** [comment]
> Is this in the preview or stable version?

Neither, I compiled that from the branch out if curiosity. It's still under active development. Let's be patient haha.

**Comment by jmcphers** [comment]
Hi all, I'm pleased to let you know that we have just completed an initial run at this feature, and it is available as of today's build (2026.04 build 4). You can opt in to the experimental feature using this setting:

<img width="900" height="119" alt="Image" src="https://github.com/user-attachments/assets/47b7e86b-e59f-4b29-ae4e-da1cd6e7c846" />

Note that this experimental version, while it does get the basics into Positron, does not have support for many popular RStudio features:

- Figure options such as `height` and `width` are not read or respected
- Data results are printed plainly, not as interactive tables
- HTML widgets from R are not rendered
- Previews are not shown for equations
- Progress (and errors) are only shown at the cell level, not line by line
- Knitr-style options are not read or respected
- Cell execution options are only read from the cell itself

More details in https://github.com/posit-dev/positron/pull/12036.

That said, if you encounter bugs or have ideas for additional enhancements (we are already tracking all of the above), please open an issue or let us know on the discussion forum.

**Comment by olalobeja** [comment]
Many thanks @jmcphers  and to the entire crew doing what you do —keep on cheerfully! Will definitely give feedback. Thanks!

**Comment by cccfran** [comment]
This is awesome! I used it for teaching already and it worked well! Students like it since most have vscode experience with Python. Thank you all so much!

Some feedback:
1. I wish to have an option to sync the kernel between the editor and the console (just like RStudio); while also keeping the option to start new kernels. I found it very convenient to test/run things temporarily. I can also leverage the plot panel to show HTML widgets. 
2. Some latex equations (use \\$$xxx$$) have "Preview" at the top left corner but some don't. Not sure what would be the preferred behavior. But hovering works pretty well.

**Comment by juliasilge** [comment]
@cccfran The consoles in general are each a separate session, much like notebook or `.qmd` sessions are separate from each other, but you can open a special "notebook console" for any given notebook or `.qmd` by opting in to the experimental `console.showNotebookConsoles` setting. This will give you a console session that is linked to each of your notebook or `.qmd` sessions (so a separate one for each).

**Comment by rodrigosf672** [comment]
## Verification/Validation Test Plan

### Core Functionality Verification

**Open .qmd/.rmd as Notebook**
- Open existing .qmd file in Positron
- Verify code chunks are recognized and executable
- Execute a chunk → Verify output appears inline below chunk
- Compare to .ipynb experience → Document differences

**Inline Output Display**
- Execute chunk with text output → Verify appears inline
- Execute chunk with plot → Verify plot renders inline
- Execute chunk with table → Verify table displays inline
- Check output positioning matches RStudio behavior

**Multiple Chunk Execution**
- Run all chunks sequentially
- Verify outputs appear for each chunk
- Check execution order is correct

### Version Control Validation

**Plain Text Format**
- View .qmd file in text editor → Verify human-readable
- Compare to .ipynb → Verify .qmd is plain text, .ipynb is JSON

**Git Diff Readability**
- Make changes to .qmd file
- Run `git diff` → Verify diff shows actual code changes clearly
- Compare with .ipynb diff → Verify .qmd is more readable

**Output Not Saved**
- Execute chunks with outputs
- Save file → Close and reopen
- Verify outputs are NOT persisted in file (unlike .ipynb)
- Re-execute to regenerate outputs

**No nbstripout Needed**
- Commit .qmd to git
- Verify no output pollution in commits
- Check workflow is simpler than .ipynb (no third-party tools)

### Language Support

**R Chunks**
- Execute R code chunk in .qmd
- Verify R kernel used and output appears

**Python Chunks**
- Execute Python code chunk in .qmd
- Verify Python kernel used and output appears

**Mixed Languages**
- .qmd with both R and Python chunks
- Execute both → Verify correct kernels used

### YAML and Markdown

**YAML Frontmatter**
- .qmd with YAML header
- Verify YAML is not executed as code
- Check metadata is preserved

**Markdown Rendering**
- Mix of markdown and code chunks
- Verify markdown sections render appropriately
- Check editing works smoothly

### Platform Testing

- Windows → Test chunk execution and inline output
- macOS → Test chunk execution and inline output  
- Linux → Test chunk execution and inline output

---

## E2E Automation Assessment

**HIGH PRIORITY for Playwright automation:**
1. ✅ Open .qmd file in Positron
2. ✅ Execute code chunk
3. ✅ Verify output appears inline
4. ✅ Save file and verify outputs not persisted
5. ✅ Reopen file and verify can re-execute

**MEDIUM PRIORITY:**
- Multiple chunk execution
- Mixed R/Python chunks
- YAML handling

**LOW PRIORITY (manual testing acceptable):**
- Git diff readability (better as manual inspection)
- Comparison with RStudio (subjective UX)
- Version control workflow benefits

**Recommended Playwright Test:**
```typescript
test('qmd notebook inline output', async ({ page }) => {
  // Open .qmd file
  // Execute first code chunk
  // Verify output appears inline
  // Save file
  // Close and reopen
  // Verify outputs not persisted
});
```

**Note:** This feature is a significant enhancement and should have comprehensive E2E coverage.

---

*Generated by Claude*

**Comment by cccfran** [comment]
Thanks, @juliasilge — it’s a very cool feature!

One thing I’ve noticed: when I use the `.rmd` console, plotting behaviors are a bit unexpected. I get a small preview/snippet in the `.rmd` console, but it doesn’t sync to the Plot pane. For HTML widgets, nothing renders at all. I was expecting them to show up in the Plot pane via the `.rmd` console.

If I plot from a Python chunk (with for example matplotlib) in an `.rmd`, the figure shows up in the Plot pane, but not inline in the document.

**Comment by juliasilge** [comment]
Yes, we consider those "notebook console" options experimental and are aware that they are not ready for prime time: https://github.com/posit-dev/positron/issues/10709

**Comment by rodrigosf672** [comment]
### Verified Fixed
    Positron Version(s) : 2026.04.0 build 93
    OS Version          : macOS Tahoe 26.3

### Test scenario(s) 
- Support has been implemented.
